modules/tensor_mechanics/src/materials/FiniteStrainUObasedCP.C

```diff
@@ -0,0 +1,851 @@
+    {
+      dtaudpk2[j] = (*_flow_direction[i])[_qp][j];
+      dfpinvdslip[j] = - _fp_old_inv * (*_flow_direction[i])[_qp][j];
+    }
+
+    for (unsigned int j = 0; j < nss; j++)
+      dfpinvdpk2 += (dfpinvdslip[j] * dslipdtau[j] * _dt).outerProduct(dtaudpk2[j]);
+  }
+  jac = RankFourTensor::IdentityFour() - (_elasticity_tensor[_qp] * deedfe * dfedfpinv * dfpinvdpk2);
+}
+
+// Calls getMatRot to perform RU factorization of a tensor.
+RankTwoTensor
+FiniteStrainUObasedCP::get_current_rotation(const RankTwoTensor & a)
+{
+  return getMatRot(a);
+}
+
+// Performs RU factorization of a tensor
+RankTwoTensor
+FiniteStrainUObasedCP::getMatRot(const RankTwoTensor & a)
+{
+  RankTwoTensor rot;
+  RankTwoTensor c, diag, evec;
+  PetscScalar cmat[LIBMESH_DIM][LIBMESH_DIM], work[10];
+  PetscReal w[LIBMESH_DIM];
+  PetscBLASInt nd = LIBMESH_DIM,
+               lwork = 10,
+               info;
+
+  c = a.transpose() * a;
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      cmat[i][j] = c(i,j);
+
+  LAPACKsyev_("V", "U", &nd, &cmat[0][0], &nd, w, work, &lwork, &info);
+
+  if (info != 0)
+    mooseError("FiniteStrainUObasedCP: DSYEV function call in getMatRot function failed");
+
+  diag.zero();
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    diag(i,i) = std::pow(w[i], 0.5);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      evec(i,j) = cmat[i][j];
+
+  rot = a * ((evec.transpose() * diag * evec).inverse());
+
+  return rot;
+}
+
+// Calculates tangent moduli which is used for global solve
+void
+FiniteStrainUObasedCP::computeQpElasticityTensor()
+{
+
+}
+
+ElasticityTensorR4
+FiniteStrainUObasedCP::calcTangentModuli()
+{
+  ElasticityTensorR4 tan_mod;
+
+  switch (_tan_mod_type)
+  {
+    case 0:
+      tan_mod = elastoPlasticTangentModuli();
+      break;
+    default:
+      tan_mod = elasticTangentModuli();
+  }
+
+  return tan_mod;
+}
+
+ElasticityTensorR4
+FiniteStrainUObasedCP::elastoPlasticTangentModuli()
+{
+  ElasticityTensorR4 tan_mod;
+  RankTwoTensor pk2fet, fepk2;
+  RankFourTensor deedfe, dsigdpk2dfe;
+
+  // Fill in the matrix stiffness material property
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+      {
+        deedfe(i,j,k,i) = deedfe(i,j,k,i) + _fe(k,j) * 0.5;
+        deedfe(i,j,k,j) = deedfe(i,j,k,j) + _fe(k,i) * 0.5;
+      }
+
+  dsigdpk2dfe = _fe.mixedProductIkJl(_fe) * _elasticity_tensor[_qp] * deedfe;
+
+  pk2fet = _pk2[_qp] * _fe.transpose();
+  fepk2 = _fe * _pk2[_qp];
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int l = 0; l < LIBMESH_DIM; ++l)
+      {
+        tan_mod(i,j,i,l) = tan_mod(i,j,i,l) + pk2fet(l,j);
+        tan_mod(i,j,j,l) = tan_mod(i,j,j,l) + fepk2(i,l);
+      }
+
+  tan_mod += dsigdpk2dfe;
```

Comment:
You have to include dfedf * dfsstretch_inc. Can check the hyperelasticviscoplastic tangent moduli,